text/0000-dockerfiles.md

```diff
@@ -0,0 +1,218 @@
+
+# Motivation
+[motivation]: #motivation
+
+[RFC0069](https://github.com/buildpacks/rfcs/blob/main/text/0069-stack-buildpacks.md) introduces complexity by defining an API that allows buildpacks to modify base images. To avoid this complexity, we could rely on generated Dockerfiles for base image manipulation. This would simplify the original proposal by, e.g., only requiring a copy of the buildpack on the build-time base image.
+
+# What it is
+[what-it-is]: #what-it-is
+
+This RFC proposes that we replace stackpacks with dynamically-generated build-time and runtime Dockerfiles that act as pre-build base image extensions.
+These extensions would participate in detection and execute before the buildpack build process.
+
+For a given application, a build that uses extensions could be optimized by creating a more narrowly-scoped builder that does not contain extensions.
+
+# How it Works
+[how-it-works]: #how-it-works
+
+Note: kaniko, buildah, BuildKit, or the original Docker daemon may be used to apply Dockerfiles at the platform's discretion. The order of operations would be something like the following:
+* analyze
+* detect
+* restore
+* <new lifecycle phase> run extensions' bin/build, output Dockerfiles are written to a volume
+* <new lifecycle phase OR platform> apply Dockerfiles to run image (could run in parallel with below two)
+* <new lifecycle phase OR platform> apply Dockerfiles to build image
+* build
+* export
+
+### Dynamically-applied Dockerfiles
+
+A builder image may include any number of "extensions" directories in `/cnb/ext/`.
+
+Extensions are similar to buildpacks: they have a `/bin/build` and `/bin/detect` executable.
+However, instead of a `buildpack.toml` file, extensions have a `extension.toml` file:
+```toml
+api = "<buildpack API version>"
+
+[extension]
+id = "<extension ID>"
+name = "<extension name>"
+version = "<extension version>"
+homepage = "<extension homepage>"
+description = "<extension description>"
+keywords = [ "<string>" ]
+
+[[extension.licenses]]
+type = "<string>"
+uri = "<uri>"
+```
+
+Extensions may be packaged and examined similar to buildpacks, but with analogous `pack extension` subcommands.
+
+Other pack CLI commands, such as `pack builder create`, will be extended to include support for extensions.
+
+Unlike buildpacks,
+- Extensions must not be included in a meta-buildpacks
+- Extensions must not have `order`/`group` definitions in `extension.toml`
+
+Extensions participate in the buildpack detection process, with the same UID, GID, and interface for `/bin/detect`.
+However,
+- `/bin/detect` is optional for extensions, and they are assumed to pass detection when it is not present. Just like with buildpacks, a /bin/detect that exits with a 0 exit code passes detection, and fails otherwise.
+- Extensions may only output `provides` entries to the build plan. They must not output `requires`.
+- Extensions must all precede regular buildpacks in `order` definitions (e.g., in `builder.toml`).
```

Comment:
@sclevine on digging into this a bit further, I'm confused as to how this would work. As I understand, extensions are not allowed to output `requires` to the build plan - they can only output `provides`. However in resolving the build plan, the lifecycle currently requires all `requires` to be [provided first](https://github.com/buildpacks/lifecycle/blob/b6408639c0b0daada589b7b59a8eda2631bc4738/detector_test.go#L856). This would imply that extensions should come **last** in the order definition.

Separately in [this comment](https://github.com/buildpacks/samples/pull/113#discussion_r726206827) it was suggested to have a separate `[[order.ext]]` table in builder.toml. How would this work, if the detector is iterating through each group in the order and choosing the first one where all non-optional buildpacks pass detect? If there are two orders, that implies two detects - in which case, extensions and buildpacks cannot collaborate on the build plan.

Reply:
From an implementation perspective, the simplest thing would be to allow extensions to appear anywhere in the order, but add an `extension = <bool>` key to [[[order.group]]](https://github.com/buildpacks/spec/blob/74f91666fc22e7568acdd0ca9e2e314ff150d740/platform.md#ordertoml-toml).

Reply:
@natalieparellano 

> However in resolving the build plan, the lifecycle currently requires all requires to be provided first.

Despite the confusing title on that test providing buildpacks need to precede requiring buildpacks in the order. For example a maven buildpacks requires a JDK to execute, so the buildpack that provides the JDK must precede the maven buildpack in the order.

> From an implementation perspective, the simplest thing would be to allow extensions to appear anywhere in the order, but add an extension = <bool> key to [[order.group]].

Detect order should match execution order. If we interleave extensions and regular buildpacks we won't be able to run the extensions in a separate phase that generates the build image prior to the build phase.

Reply:
We also discussed with @sclevine how the `[[order-ext]]` table could be prepended to the detect order as a meta buildpack, allowing buildpacks and extensions to collaborate on the build plan while keeping the tables separate.

Reply:
Marking this as resolved and added a code suggestion to make it clearer.